# Prompting with Gemini
Notebook for experimenting different promting techniques with Gemini

Install the required dependencies with the following command.

In [1]:
%pip install google-generativeai

Note: you may need to restart the kernel to use updated packages.


Import the required libraries.

In [2]:
import google.generativeai as genai
import os

Insert your Gemini API key to the following command and import the Gemini model.

In [ ]:
# API_KEY = os.environ.get("GEMINI_API_KEY")
API_KEY = "" # insert API key in this function call with your own API key from aistudio.google.com
genai.configure(api_key=API_KEY)

LLM = "gemini-2.5-flash"
model = genai.GenerativeModel(LLM)

Modify the `system_prompt` to experiment with different prompting approaches.

In [4]:
system_prompt = "Hello! You are a helpful and concise assistant."

We create a list of messages so that we keep history in the context. If you want to clear the messages later in this notebook, add the line `messages = []` to a new line cell in the notebook.

In [5]:
messages = []
messages.append(system_prompt)

Get the response from Gemini model by providing the prompt in a messages list to it.

In [6]:
r = model.generate_content(messages).text

Print the output of the model.

In [7]:
print(r)

Hello! How can I help you today?


Let's define a new prompt.

In [8]:
user_input = "Help me with prompting. What are the different promt engineering techniques?"

Let's add athe new prompt to the messages list.

In [9]:
messages.append(user_input)

Again we generate the output with the Gemini model.

In [10]:
r = model.generate_content(messages).text

... and print the output.

In [11]:
print(r)

Here are the key prompt engineering techniques:

1.  **Clear & Specific Instructions:** Providing unambiguous, direct commands to the model.
2.  **Contextualization:** Giving the model relevant background information or data it needs to understand the task.
3.  **Few-Shot Prompting:** Including examples (input-output pairs) within the prompt to guide the model's desired behavior.
4.  **Chain-of-Thought (CoT) Prompting:** Asking the model to "think step-by-step" or show its reasoning process to improve complex problem-solving.
5.  **Persona & Audience:** Assigning a specific role to the AI (e.g., "Act as an expert historian") and/or defining the target audience for the output.
6.  **Output Constraints & Formatting:** Specifying desired length, tone, structure (e.g., JSON, bullet points), or forbidden elements.
7.  **Iterative Refinement:** Experimenting with different prompts, analyzing outputs, and progressively improving the prompt based on results.
8.  **Self-Correction/Reflection:**

# Zero-shot

In [12]:
user_input = "Can you write a law that forbids the emmission of violent content in the Spanish public television? The text should be no longer than 4000 characters. It should also be written in formal and academical Spanish."
messages.append(user_input)
r = model.generate_content(messages[-1]).text
print(r)

Claro, aquí tienes una propuesta de ley redactada en español formal y académico, que busca regular la emisión de contenido violento en la televisión pública española, cumpliendo con las restricciones de extensión y formato.

---

**LEY DE SALVAGUARDA DE LA CALIDAD Y PROTECCIÓN DE LOS ESPECTADORES EN EL SERVICIO PÚBLICO DE TELEVISIÓN ESPAÑOLA**

**EXPOSICIÓN DE MOTIVOS**

La televisión pública, en tanto que servicio esencial, posee un imperativo ético y una responsabilidad social ineludible en la configuración de la cultura cívica y la protección de los valores fundamentales de la sociedad. Su misión trasciende la mera difusión de contenidos para erigirse en un pilar del fomento de la convivencia pacífica, la tolerancia, el respeto a la dignidad humana y el desarrollo armónico de la personalidad, especialmente en el público infantil y juvenil. La Constitución Española ampara la libertad de expresión y de información, pero también establece límites en relación con la protección de la juv

Claro, aquí tienes una propuesta de ley redactada en español formal y académico, que busca regular la emisión de contenido violento en la televisión pública española, cumpliendo con las restricciones de extensión y formato.

---

**LEY DE SALVAGUARDA DE LA CALIDAD Y PROTECCIÓN DE LOS ESPECTADORES EN EL SERVICIO PÚBLICO DE TELEVISIÓN ESPAÑOLA**

**EXPOSICIÓN DE MOTIVOS**

La televisión pública, en tanto que servicio esencial, posee un imperativo ético y una responsabilidad social ineludible en la configuración de la cultura cívica y la protección de los valores fundamentales de la sociedad. Su misión trasciende la mera difusión de contenidos para erigirse en un pilar del fomento de la convivencia pacífica, la tolerancia, el respeto a la dignidad humana y el desarrollo armónico de la personalidad, especialmente en el público infantil y juvenil. La Constitución Española ampara la libertad de expresión y de información, pero también establece límites en relación con la protección de la juventud y la infancia. En este contexto, se hace imprescindible establecer un marco regulatorio que module la emisión de representaciones violentas, evitando su glorificación o trivialización y garantizando que el servicio público de televisión sea un espacio seguro y enriquecedor para todos los ciudadanos. La presente ley persigue, por tanto, conciliar el derecho a la información y a la creación artística con la inexcusable obligación de tutelar el interés superior de los menores y de promover una sociedad libre de violencia.

**Artículo 1. Objeto**

La presente ley tiene por objeto establecer los principios y directrices para la regulación y supervisión de los contenidos audiovisuales que representen violencia en los servicios de televisión de titularidad pública en el ámbito estatal y autonómico, con el fin primordial de proteger a los espectadores, en especial a los menores de edad, y salvaguardar la calidad y los valores intrínsecos al servicio público.

**Artículo 2. Ámbito de Aplicación**

Esta normativa será de aplicación a todos los prestadores del servicio público de televisión de ámbito estatal (Corporación RTVE) y autonómico, así como a las sociedades filiales y cualesquiera otras entidades que gestionen canales o plataformas bajo su titularidad o control, en relación con la totalidad de los contenidos audiovisuales que emitan.

**Artículo 3. Principios Rectores de Contenidos**

Los prestadores del servicio público de televisión observarán los siguientes principios en la programación y emisión de contenidos:
a) No glorificación, apología o trivialización de la violencia, la crueldad, el terrorismo o cualquier forma de discriminación o incitación al odio.
b) Promoción de la cultura de paz, el diálogo, la resolución pacífica de conflictos y el respeto a los derechos humanos y a la dignidad de la persona.
c) Tutela reforzada de los menores, evitando la exposición a contenidos que puedan dañar su desarrollo físico, mental, moral o social.
d) Rigor, veracidad y contextualización en el tratamiento de la violencia, cuando su abordaje sea imprescindible por razones informativas, históricas, culturales o artísticas.

**Artículo 4. Prohibiciones Específicas**

Se prohíbe la emisión de contenidos audiovisuales que, de forma gratuita, injustificada o sensacionalista, incurran en los siguientes supuestos:
a) Representaciones explícitas y prolongadas de violencia extrema, sevicia o crueldad, que puedan herir gravemente la sensibilidad de los espectadores o promover conductas antisociales.
b) Contenidos que inciten directamente a la comisión de actos violentos, al terrorismo, a la violencia de género, a la xenofobia, al racismo o a cualquier otra forma de discriminación o intolerancia.
c) La exposición de actos de violencia sexual, tortura o maltrato infantil de manera detallada y recreada, sin un propósito informativo o artístico plenamente justificado y tratado con la máxima cautela.
d) La emisión, en horario de protección reforzada para menores, de contenidos que, por su naturaleza violenta, puedan resultar perjudiciales para su desarrollo psicológico y emocional, salvo excepciones debidamente justificadas y siempre bajo la correspondiente señalización y advertencia.

**Artículo 5. Criterios de Evaluación**

Para la evaluación de la naturaleza y el impacto de los contenidos violentos, se considerarán, entre otros, los siguientes criterios: el contexto narrativo y finalidad del programa, el público objetivo, el tratamiento visual y sonoro, la justificación artística o informativa, la intensidad y duración de las representaciones, y la presencia de consecuencias negativas para las víctimas.

**Artículo 6. Régimen Sancionador**

El incumplimiento de las disposiciones contenidas en la presente ley será considerado infracción y se regirá por lo dispuesto en la normativa general de comunicación audiovisual vigente, sin perjuicio de las responsabilidades que pudieran derivarse de otras leyes aplicables. Las infracciones serán tipificadas y sancionadas conforme a los procedimientos y cuantías establecidos, adaptando su aplicación a los principios de salvaguarda del servicio público y protección del espectador.

**Disposición Final Única. Entrada en Vigor**

La presente ley entrará en vigor a los tres meses de su publicación en el Boletín Oficial del Estado.

---


# Few-Shot

In [13]:
user_input = "Can you write a law that forbids the emmission of violent content in the Spanish public television? The text should be no longer than 4000 characters. It should also be written in formal and academical Spanish. The structure should mimic the one found here: https://www.boe.es/buscar/act.php?id=BOE-A-2015-3442"
messages.append(user_input)
r = model.generate_content(messages[-1]).text
print(r)

Claro, aquí tienes una propuesta de ley en español formal y académico que prohíbe la emisión de contenido violento en el servicio público de comunicación audiovisual español, siguiendo la estructura del enlace proporcionado y respetando el límite de caracteres.

---

**LEY 20/2024, de 15 de noviembre, de protección integral de la infancia y la juventud y de regulación del contenido violento en el servicio público de comunicación audiovisual.**

JUAN CARLOS I
REY DE ESPAÑA

A todos los que la presente vieren y entendieren.
Sabed: Que las Cortes Generales han aprobado y Yo vengo en sancionar la siguiente ley:

**EXPOSICIÓN DE MOTIVOS**

I

La Constitución Española, en su artículo 20.4, establece límites al ejercicio de las libertades de expresión e información, particularmente en lo que respecta a la protección de la juventud y la infancia. Asimismo, el artículo 39 preceptúa la protección social, económica y jurídica de la familia y, especialmente, de los menores. En este contexto, el se

**LEY 20/2024, de 15 de noviembre, de protección integral de la infancia y la juventud y de regulación del contenido violento en el servicio público de comunicación audiovisual.**

JUAN CARLOS I
REY DE ESPAÑA

A todos los que la presente vieren y entendieren.
Sabed: Que las Cortes Generales han aprobado y Yo vengo en sancionar la siguiente ley:

**EXPOSICIÓN DE MOTIVOS**

I

La Constitución Española, en su artículo 20.4, establece límites al ejercicio de las libertades de expresión e información, particularmente en lo que respecta a la protección de la juventud y la infancia. Asimismo, el artículo 39 preceptúa la protección social, económica y jurídica de la familia y, especialmente, de los menores. En este contexto, el servicio público de comunicación audiovisual desempeña un papel fundamental como vector de cultura, educación y cohesión social, debiendo, por tanto, ser garante de los principios y valores democráticos, entre los que destaca el respeto a la dignidad humana y la erradicación de la violencia.

La exposición a contenidos violentos, especialmente para los segmentos de población más vulnerables como son los menores, ha sido identificada como un factor que puede incidir negativamente en su desarrollo psicológico y social. Diversos estudios y convenios internacionales, como la Convención sobre los Derechos del Niño, ratifican la necesidad imperativa de proteger a la infancia y la juventud de toda forma de maltrato o explotación, incluyendo la exposición a contenidos que puedan incitar a la violencia, glorificarla o normalizarla.

II

Esta ley tiene como objeto principal establecer un marco normativo claro y robusto que prohíba la emisión de contenido violento de carácter gratuito, explícito o que incite a la violencia, el odio o la discriminación en los prestadores del servicio público de comunicación audiovisual de ámbito estatal y autonómico. Se busca así asegurar que dichos servicios cumplan con su función social y educativa, promoviendo valores de paz, tolerancia y respeto.

La presente ley se estructura en un Título Preliminar, que define el objeto y el ámbito de aplicación, y dos Títulos que abordan la prohibición y regulación del contenido violento, y el régimen sancionador, respectivamente.

**TÍTULO PRELIMINAR**
**Disposiciones generales**

**Artículo 1. Objeto.**
La presente ley tiene por objeto la protección integral de la infancia y la juventud y la salvaguarda de la dignidad humana mediante la prohibición y regulación de la emisión de contenido violento en el servicio público de comunicación audiovisual.

**Artículo 2. Ámbito de aplicación.**
Esta ley será de aplicación a todos los prestadores del servicio público de comunicación audiovisual de titularidad estatal y autonómica, así como a las sociedades mercantiles gestoras de dichos servicios.

**Artículo 3. Definiciones.**
A los efectos de esta ley, se entenderá por:
a) Contenido violento: toda representación audiovisual que muestre actos de agresión física, psicológica o sexual, autodestrucción, crueldad o sufrimiento, y que tenga un carácter gratuito, explícito, excesivo, vejatorio, que incite al odio, a la discriminación, a la comisión de delitos o que glorifique la violencia como solución a conflictos, sin un fin informativo, artístico, educativo o de denuncia justificado y contextualizado.

**TÍTULO I**
**Prohibición y regulación del contenido violento**

**Artículo 4. Prohibición de contenido violento.**
1. Se prohíbe la emisión, bajo cualquier modalidad, de contenido violento conforme a la definición establecida en el artículo 3.a) de esta ley, por parte de los prestadores del servicio público de comunicación audiovisual.
2. La prohibición abarca, con carácter enunciativo pero no limitativo, aquellos contenidos que:
a) Exalten o promuevan la violencia de género, intrafamiliar, sexual, terrorista o de cualquier otra índole.
b) Muestren actos de crueldad extrema o tortura de forma gratuita.
c) Inciten directamente al odio o la discriminación por razón de nacimiento, raza, sexo, religión, opinión o cualquier otra condición o circunstancia personal o social.
d) Puedan causar grave perjuicio psíquico o moral a los menores, aun cuando no se trate de una incitación directa.

**Artículo 5. Criterios de evaluación.**
La evaluación del carácter violento de un contenido deberá realizarse atendiendo al contexto global de la obra, su finalidad, el tratamiento narrativo, la intensidad y duración de las escenas, y el mensaje final que se transmita, así como su potencial impacto en la audiencia, especialmente en los menores.

**Artículo 6. Obligaciones de los prestadores del servicio público de comunicación audiovisual.**
1. Los prestadores deberán establecer códigos de conducta y protocolos internos que garanticen el cumplimiento de lo dispuesto en esta ley.
2. Promoverán la formación y sensibilización de sus profesionales en la prevención de la emisión de contenido violento.
3. Adoptarán medidas técnicas y organizativas para asegurar la verificación previa de todo el contenido antes de su emisión.

**TÍTULO II**
**Régimen sancionador**

**Artículo 7. Infracciones.**
El incumplimiento de las prohibiciones y obligaciones establecidas en la presente ley será considerado infracción y se graduará en leves, graves y muy graves, conforme a lo establecido en la Ley 7/2010, de 31 de marzo, General de la Comunicación Audiovisual, y demás normativa aplicable.

**Artículo 8. Sanciones.**
Las infracciones tipificadas en el artículo anterior serán sancionadas de acuerdo con el régimen sancionador previsto en la Ley 7/2010, de 31 de marzo, General de la Comunicación Audiovisual, sin perjuicio de las responsabilidades civiles o penales que pudieran derivarse.

**DISPOSICIÓN ADICIONAL ÚNICA. Adaptación normativa.**
Los prestadores del servicio público de comunicación audiovisual deberán adaptar sus códigos de conducta, protocolos internos y manuales de estilo a lo dispuesto en la presente ley en el plazo de seis meses desde su entrada en vigor.

**DISPOSICIÓN DEROGATORIA ÚNICA. Derogación normativa.**
Quedan derogadas cuantas disposiciones de igual o inferior rango se opongan a lo establecido en la presente ley.

**DISPOSICIÓN FINAL PRIMERA. Título competencial.**
La presente ley se dicta al amparo de la competencia exclusiva del Estado en materia de legislación básica sobre contratos y en desarrollo de los derechos fundamentales y libertades públicas, de conformidad con lo previsto en el artículo 149.1.1ª y 149.1.27ª de la Constitución Española.

**DISPOSICIÓN FINAL SEGUNDA. Entrada en vigor.**
La presente ley entrará en vigor el día siguiente al de su publicación en el «Boletín Oficial del Estado».

Por tanto,
Mando a todos los españoles, particulares y autoridades, que guarden y hagan guardar esta ley.

Madrid, 15 de noviembre de 2024.

JUAN CARLOS I

El Presidente del Gobierno,
[Nombre del Presidente del Gobierno]

# Chain of thought

In [14]:
user_input = """
Can you write a law that forbids the emmission of violent content in the Spanish public television? The text should be no longer than 4000 characters.
It should also be written in formal and academical Spanish. The structure should mimic the one found here: https://www.boe.es/buscar/act.php?id=BOE-A-2015-3442.
Expose some reasons that led to this legislation in the preamble. Some articles should refer to these reasons by explaining how they affect them.
Replace al the placeholders with actual values, for example, the law should be written using today's date and signed by the current president and king of Spain. Make up the law number.
"""
messages.append(user_input)
r = model.generate_content(messages[-1]).text
print(r)

**LEY 12/2024, DE 10 DE JUNIO, DE PROTECCIÓN DE LA AUDIENCIA FRENTE AL CONTENIDO VIOLENTO EN LOS SERVICIOS DE TELEVISIÓN PÚBLICA.**

**PREÁMBULO**

La presente Ley responde a la imperativa necesidad de salvaguardar la integridad psicosocial de la ciudadanía, y de manera preeminente la de la infancia y la juventud, frente a la exposición indiscriminada a contenidos violentos en los servicios de comunicación audiovisual de titularidad pública. La evidencia científica y la experiencia social demuestran que la exposición reiterada a la violencia gráfica o simbólica puede generar efectos adversos significativos, tales como la desensibilización, el incremento de la ansiedad, la promoción de conductas agresivas y la trivialización del sufrimiento humano. En este contexto, el servicio público de radiodifusión, sustentado por fondos públicos y concebido como herramienta para la cohesión social y la promoción de valores democráticos, posee la misión esencial de fomentar una cultura de paz, respe

**LEY 12/2024, DE 10 DE JUNIO, DE PROTECCIÓN DE LA AUDIENCIA FRENTE AL CONTENIDO VIOLENTO EN LOS SERVICIOS DE TELEVISIÓN PÚBLICA.**

**PREÁMBULO**

La presente Ley responde a la imperativa necesidad de salvaguardar la integridad psicosocial de la ciudadanía, y de manera preeminente la de la infancia y la juventud, frente a la exposición indiscriminada a contenidos violentos en los servicios de comunicación audiovisual de titularidad pública. La evidencia científica y la experiencia social demuestran que la exposición reiterada a la violencia gráfica o simbólica puede generar efectos adversos significativos, tales como la desensibilización, el incremento de la ansiedad, la promoción de conductas agresivas y la trivialización del sufrimiento humano. En este contexto, el servicio público de radiodifusión, sustentado por fondos públicos y concebido como herramienta para la cohesión social y la promoción de valores democráticos, posee la misión esencial de fomentar una cultura de paz, respeto mutuo y convivencia cívica, alejada de la glorificación, incitación o normalización de la violencia. Esta Ley, por consiguiente, refuerza el compromiso del Estado con el bienestar emocional colectivo, el desarrollo armónico de las nuevas generaciones y la calidad ética del espacio audiovisual público.

**TÍTULO PRELIMINAR. DISPOSICIONES GENERALES**

Artículo 1. Objeto.
La presente Ley tiene por objeto establecer un marco normativo que prohíba la emisión de contenido violento en los servicios de televisión operados por entidades del sector público.

Artículo 2. Ámbito de aplicación.
Esta Ley será de aplicación a todos los prestadores de servicios de comunicación audiovisual televisiva cuya titularidad corresponda al Estado, a las Comunidades Autónomas o a las Corporaciones Locales, así como a sus entidades participadas mayoritariamente.

Artículo 3. Definición de contenido violento.
A los efectos de esta Ley, se entenderá por contenido violento toda representación audiovisual que, por su crudeza, explicitud, intensidad o contexto, muestre actos de agresión física o psicológica extrema, crueldad manifiesta, mutilación, tortura, abuso, o que glorifique, incite o trivialice la violencia, el odio, la discriminación o el conflicto armado, de manera que pueda perturbar gravemente el desarrollo psicosocial de los menores, inducir a la imitación de conductas peligrosas o generar angustia injustificada en la audiencia general, contraviniendo los principios de dignidad humana y convivencia pacífica.

**TÍTULO I. DE LA PROHIBICIÓN Y SUS CRITERIOS DE EVALUACIÓN**

Artículo 4. Prohibición general.
Se prohíbe la emisión de cualquier contenido que se ajuste a la definición de contenido violento establecida en el Artículo 3 en la programación de los servicios de televisión de ámbito público.

Artículo 5. Criterios de evaluación.
La evaluación de la naturaleza violenta de un contenido se realizará atendiendo, entre otros factores, a su contexto narrativo, la veracidad del hecho representado, su finalidad didáctica o informativa, el horario de emisión, y muy especialmente, a la potencialidad de generar los impactos negativos descritos en el preámbulo, con particular énfasis en la protección de los menores.

**TÍTULO II. DE LA SUPERVISIÓN Y CONTROL**

Artículo 6. Autoridad competente.
La Comisión Nacional de los Mercados y la Competencia (CNMC) será el organismo competente para velar por el estricto cumplimiento de las disposiciones de esta Ley, estableciendo los procedimientos de denuncia, inspección y régimen sancionador que correspondan.

**DISPOSICIÓN DEROGATORIA ÚNICA.**
Quedan derogadas cuantas disposiciones de igual o inferior rango se opongan a lo establecido en la presente Ley.

**DISPOSICIONES FINALES**

Primera. Habilitación normativa.
Se habilita al Gobierno para dictar cuantas disposiciones reglamentarias sean necesarias para el desarrollo y la eficaz aplicación de la presente Ley.

Segunda. Entrada en vigor.
La presente Ley entrará en vigor a los seis meses de su publicación en el Boletín Oficial del Estado.

Por tanto,
Mando a todos los españoles, particulares y autoridades, que guarden y hagan guardar esta Ley.

Madrid, 10 de junio de 2024.

FELIPE R.

El Presidente del Gobierno,
PEDRO SÁNCHEZ PÉREZ-CASTEJÓN
